File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) 
    {
        int carry = 0;
        ListNode ans = new ListNode();
        ListNode top = ans;

        while(l1!=null || l2!=null)
        {
            int val1 = (l1==null ? 0 : l1.val);
            int val2 = (l2==null ? 0 : l2.val);

            int sum = val1 + val2 + carry;
            carry = (sum > 9 ? 1 : 0);
            if(sum > 9) sum = sum-10; 

            ListNode n = new ListNode(sum);
            top.next = n;
            top = n;

            if(l1 != null) l1 = l1.next;
            if(l2 != null) l2 = l2.next;
        }   

        if(carry == 1)
        {
            ListNode n = new ListNode(1);
            top.next = n;
            top = n;
        }   

        return ans.next;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* add(int carry, ListNode* l1, ListNode* l2){
        if(!l1 && !l2) {
            if(carry == 0) return nullptr;
            return new ListNode(carry);
        }
        int sum = carry + (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
        carry = sum/10;
        sum = sum%10;
        ListNode* ans = new ListNode(0);
        ans->val = sum;
        ans->next = add(carry, (l1 ? l1->next : nullptr), (l2 ? l2->next : nullptr));
        return ans;
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = add(0, l1, l2);
        return head;
    }
};